python/packages/agent_framework.amazon/agent_framework_bedrock/_chat_client.py

```diff
@@ -0,0 +1,504 @@
+            raise ServiceInitializationError("Failed to initialize Bedrock settings.", ex) from ex
+
+        if bedrock_runtime_client is None:
+            session = boto3_session or self._create_session(settings)
+            bedrock_runtime_client = session.client(
+                "bedrock-runtime",
+                region_name=settings.region,
+                config=BotoConfig(user_agent_extra=AGENT_FRAMEWORK_USER_AGENT),
+            )
+
+        super().__init__(**kwargs)
+        self._bedrock_client = bedrock_runtime_client
+        self.model_id = settings.chat_model_id
+        self.region = settings.region
+
+    @staticmethod
+    def _create_session(settings: BedrockSettings) -> Boto3Session:
+        session_kwargs: dict[str, Any] = {"region_name": settings.region or DEFAULT_REGION}
+        if settings.access_key and settings.secret_key:
+            session_kwargs["aws_access_key_id"] = settings.access_key.get_secret_value()
+            session_kwargs["aws_secret_access_key"] = settings.secret_key.get_secret_value()
+        if settings.session_token:
+            session_kwargs["aws_session_token"] = settings.session_token.get_secret_value()
+        return Boto3Session(**session_kwargs)
+
+    async def _inner_get_response(
+        self,
+        *,
+        messages: MutableSequence[ChatMessage],
+        chat_options: ChatOptions,
+        **kwargs: Any,
+    ) -> ChatResponse:
+        request = self._build_converse_request(messages, chat_options, **kwargs)
+        raw_response = await asyncio.to_thread(self._bedrock_client.converse, **request)
+        return self._process_converse_response(raw_response)
+
+    async def _inner_get_streaming_response(
+        self,
+        *,
+        messages: MutableSequence[ChatMessage],
+        chat_options: ChatOptions,
+        **kwargs: Any,
+    ) -> AsyncIterable[ChatResponseUpdate]:
+        response = await self._inner_get_response(messages=messages, chat_options=chat_options, **kwargs)
+        contents = list(response.messages[0].contents if response.messages else [])
+        if response.usage_details:
+            contents.append(UsageContent(details=response.usage_details))
+        yield ChatResponseUpdate(
+            response_id=response.response_id,
+            contents=contents,
+            model_id=response.model_id,
+            finish_reason=response.finish_reason,
+            raw_response=response.raw_response,
+        )
+
+    def _build_converse_request(
+        self,
+        messages: MutableSequence[ChatMessage],
+        chat_options: ChatOptions,
+        **kwargs: Any,
+    ) -> dict[str, Any]:
+        model_id = chat_options.model_id or self.model_id
+        if not model_id:
+            raise ServiceInitializationError(
+                "Bedrock model_id is required. Set via chat options or BEDROCK_CHAT_MODEL_ID environment variable."
+            )
+
+        system_prompts = self._extract_system_prompts(messages)
```

Comment:
in these two functions, we are looping through the messages twice, maybe combine and return the first part of the payload